openllm-core/src/openllm_core/prompt/prompt_template.py

```diff
@@ -0,0 +1,42 @@
+from __future__ import annotations
+import typing as t
+
+import attr
+
+from .utils import PromptFormatter
+
+default_formatter = PromptFormatter()
+
+@attr.define(slots=True)
+class PromptTemplate:
+  _template: str | None
+  _system_message: str | None
+
+  def __init__(self, template: str | None = None, system_message: str | None = None) -> None:
+    self._template = template
+    self._system_message = system_message
+
+  def _fill_in_variables(self, **kwargs: t.Any) -> str | None:
+    if self._template is None: return self._template
+    template_variables = default_formatter.extract_template_variables(self._template)
+    prompt_variables = {key: '{' + key + '}' if key not in kwargs else (kwargs[key] or '') for key in template_variables}
+    return self._template.format(**prompt_variables)
+
+  @property
+  def template(self) -> str | None:
+    return self._fill_in_variables(system_message=self._system_message)
```

Comment:
This is confusing imo. If you want to return the fill in template, then the `_template` is placeholder, rather than template.

Reply:
Probably better to use https://www.attrs.org/en/stable/init.html#post-init and init to default template attribute.

This is also incorrect imo. All attributes in PromptTemplate should be public (at least for template and system_message).

Reply:
Think from the users perspective as well. they want to access this attribute.

nit: all attributes in attrs class are immutable, so you need to create a `with_options` or `with_prompt` or `with_system_message` for builder pattern.

Reply:
I think the big design problem here is the template does not have all the input variables at the exact same time (eg. instruction) is entered last in our current work flow. You could argue that we instantiate a PromptTemplate once we have all the inputs, but that defeats the purpose of this class entirely imo.

Reply:
Assuming that we want to have a flexible PromptTemplate class where multiple custom inputs could be added in the future, I'm thinking it might be better to not hard code in `system_message` as an attribute but rather have `input_variables` of sort. I will make that change

Reply:
This is irrelevant right? It doesn't matter if you have all of the variable for format. The purpose of the template is just to fill out the system_message, or at least this is what you have right now. 

wrt the lifecycle of the PromptTemplate, I think template should just be formatted with system_message directly after `post_init`, and users can swap out system_message if needed.

Reply:
> wrt the lifecycle of the PromptTemplate, I think template should just be formatted with system_message directly after post_init, and users can swap out system_message if needed.

I'm a bit confused with this point. I could fill in the template with the system_message with post_init but then the user wouldn't be able to swap out system_message because:
1. attributes are immutable
2. after format, the `{system_message}` in the template would be replaced and lost

By 'swap out', do you mean the user could create another instance of PromptTemplate? In that case, I am sort of lost as to what the purpose of the PromptTemplate class is. Since the user could just as easily write out the entire prepared/filled-in template as a string instead of using the class.

Reply:
Ah I see. I was thinking about this use case:

```python

ptemplate = openllm.PromptTemplate(system_message="Hello", template="<|SYSTEM|>{system_message}<|USER|>{instruction}<|ASSISTANT|>")

ptemplate.template -> "<|SYSTEM|>Hello<|USER|>{instruction}<|ASSISTANT|>"
ptemplate.with_system_message("New system message") -> "<|SYSTEM|>New system message<|USER|>{instruction}<|ASSISTANT|>"
```

btw you can always mutate attributes in python with `object.__setattr__` (also attrs has a function called `attr.evolve`)

Reply:
ah gotcha! Ok, I just implemented a new method called `with_input_variables()` which should do the above.